src/de/deserializer.zig

```diff
@@ -2,36 +2,64 @@ const concepts = @import("concepts");
 const getty = @import("getty");
 const std = @import("std");
 
-pub fn Deserializer(comptime user_dbt: anytype) type {
+pub fn Deserializer(comptime user_dbt: anytype, comptime Reader: type) type {
+    const JsonReader = std.json.Reader(1024 * 4, Reader);
     return struct {
-        allocator: ?std.mem.Allocator = null,
-        tokens: std.json.TokenStream,
+        tokens: JsonReader,
 
         const Self = @This();
 
-        pub fn init(slice: []const u8) Self {
+        pub fn init(allocator: std.mem.Allocator, reader: Reader) Self {
             return Self{
-                .tokens = std.json.TokenStream.init(slice),
+                .tokens = JsonReader.init(allocator, reader),
             };
         }
 
-        pub fn withAllocator(allocator: std.mem.Allocator, slice: []const u8) Self {
-            return Self{
-                .allocator = allocator,
-                .tokens = std.json.TokenStream.init(slice),
-            };
+        pub fn deinit(self: *Self) void {
+            self.tokens.deinit();
+            self.* = undefined;
         }
 
         /// Validates that the input data has been fully deserialized.
         ///
         /// This method should always be called after a value has been fully
         /// deserialized.
         pub fn end(self: *Self) Error!void {
-            if (try self.tokens.next()) |_| {
-                return error.InvalidTopLevelTrailing;
+            if (try self.tokens.next() != .end_of_document) {
+                return error.SyntaxError;
+            }
+        }
+
+        /// A small helper to retrieve this struct's stored allocator.
+        pub inline fn getAllocator(self: *const Self) std.mem.Allocator {
+            return self.tokens.scanner.stack.bytes.allocator;
+        }
```

Comment:
Actually, do we even need to store the allocator as a field?

Don't the deserialize methods take an allocator? Why not use that instead of the BitStack's or the one passed to `init`?

Reply:
_From Discord..._

> about getAllocator, it's currently used in a few functions where I don't have any other allocator:
> - freeToken
> - fromDeserializer

Yeah, we do need to have an allocator lying around for `fromDeserializer`. That's fine.

> having another field would be storing the allocator twice, which is kind of wasteful. you also see the pattern of taking an allocator from a child data structure a lot in std.
> 
> i do get your point about that comment in std's json stating the allocator is only used for tracking nesting, but then again that doesn't really mean we can't use it for anything else.

It's fine if we store the allocator twice; it's just two pointers. We can easily optimize this whenever we want if it turns out to be an issue.

I don't want to have to rely on `BitStack`'s implementation to grab our allocator. For example, if its `bytes` field changes from an `ArrayList` or is removed or whatever, then we'd probably have to change our code as well.

Also, even though `BitStack`'s allocator is exact same as the one passed to `init`, I just think it's confusing and messy to have to use `self.tokens.scanner.stack.bytes.allocator` just to get our allocator that was passed into `init`.

> Also, I currently use the allocator passed to the deserialize functions only for stuff that is returned from the function in case it seemed random which one i use

Let's always use the allocator passed to the `deserialize` methods (and whatever other functions take them). If `allocator == null`, we can return `error.MissingAllocator`.

The new `allocator` field in `Deserializer` will just be for `fromDeserializer` calls.